### PR TITLE
Add a Docker build target for a "scoped" variant

### PR DESCRIPTION
**Pull request by mpdude**
As an attempt to improve on #5593, here is an extension of the `Dockerfile` to include a "scoped Rector" build stage. Additionally, the PHP version to build upon can be passed as a build  argument.

```
docker build --target rector -t rector-php8.0 .
docker build --target rector-scoped -t rector-scoped-php8.0 .
docker build --target rector --build-arg PHP_VERSION=7.4 -t rector-php7.4 .
docker build --target rector-scoped --build-arg PHP_VERSION=7.4 -t rector-scoped-php7.4 .
```

A full, maintainable solution would need to integrate with GH Actions and possibly remove redundancy: Right now, there is some code from `./build-rector-scoped.sh` duplicated in the Dockerfile. To me, it seemed to be more natural this way, keeping the build process entirely in Docker and the Dockerfile.

After the Docker image has been built, the "scoped" Rector version can be copied out of it:

```
docker create -ti --name dummy rector-scoped-php8.0 bash ; docker cp dummy:/rector ./rector-scoped ; docker rm -f dummy
```

I'd be glad if this at least could serve as a starting point for a discussion on how to approach the issue of having a "scoped" Docker image.

### CODE REVIEW

File: Dockerfile
Comment (mpdude) [inline Dockerfile:6]: I had the impression we need those as runtime requirements; it that really the case?

**Comment by mpdude** [comment]
@JanMikes if you think that in general this is heading in the right direction, I could have a look at the GHA workflows and see if I get it all integrated.

**Comment by JanMikes** [comment]
Hi @mpdude there is little refactoring planned: creating separate repository for base images, so we save time in CI + make Dockerfiles a little cleaner. 

We had a call with @TomasVotruba and we agreed on following:
- Official docker images will contain scoped (prefixed) version only
- We will build images for multiple php versions - 7.3, 7.4 and 8.0

**Comment by mpdude** [comment]
That's great news!

What exactly do you mean with "separate repository"?

A repo just for the Docker-related things? Can/should I try to help elsewhere?

**Comment by mpdude** [comment]
Oh, and while we’re at it, maybe use `-alpine` base images for much smaller Docker images? 🌱

**Comment by JanMikes** [comment]
Regards alpine - im not sure if anything changes, see this comment https://github.com/rectorphp/rector/pull/3374#issuecomment-653822476

> A repo just for the Docker-related things? Can/should I try to help elsewhere?

Exactly, but not for everything, just for base images, that will contain system dependencies, php extension and so on. Specific builds will remain in this repository, so we will have something like:
```
FROM rector/php:${PHP_VERSION}
```

> Can/should I try to help elsewhere?

Ofc you can, currently im afraid it will get blocked by me for 2 more days before i get to it :(.

**Comment by mpdude** [comment]
@JanMikes ok, let's put Alpine aside for the time being.

Since I wanted to find out for myself anyway how builds could best be cached in GHA, I've opened #5667. This uses the GitHub Container Registry and the `buildx` builder to cache builds, including intermediate layers for multistage builds.

I don't know if this supersedes the idea of having dedicated base images, but we could this discuss over there.

Looking forward to your feedback!

File: Dockerfile
Comment (JanMikes) [reply Dockerfile:6]: Honestly, im not sure, they must be there for reason 😄

File: Dockerfile
Comment (JanMikes) [inline Dockerfile:16]: Lets keep it like this, but in future it is always great to keep refactoring for separate MR than adding new feature so it is easier to review.

File: Dockerfile
Comment (JanMikes) [inline Dockerfile:89]: Secured rector can use scoped version as well.

File: Dockerfile
Comment (JanMikes) [inline Dockerfile:57]: Im not sure why do we need this extra `runtime` target, why it is not directly `FROM base as rector` ?

File: Dockerfile
Comment (JanMikes) [inline Dockerfile:79]: We can use scoped only in docker

**Review by JanMikes — CHANGES_REQUESTED** [review]
Please update dockerfile so there is always and only scoped version used

**Comment by JanMikes** [comment]
> To me, it seemed to be more natural this way, keeping the build process entirely in Docker and the Dockerfile.

I like the idea, though i am almost 100 % sure @TomasVotruba will disagree as he runs it locally without docker 😄 so i would reverse it -> in docker run the bash script instead of copy paste could do the job.

**Comment by TomasVotruba** [comment]
@JanMikes If this output is preserved - https://github.com/rectorphp/rector-prefixed - I don't mind :)

**Comment by mpdude** [comment]
#5667 addresses a few of your comments already, so let's wait here until/if that gets merged and proceed with the remaining items then.

As to running the scoping thing inside Docker: I did this way and without the extra shell script because that contains some steps not really necessary in Docker; for example, copying the files around using `rsync` to yet another workdir, scope from there to some other directory, only to copy back once again... And also, we'd have to install rsync in the `build` layer just to copy the files.

But if you'd like to have an integrated approach with the same shell script used stand-alone and inside the Dockerfile, I am sure I can come up with a solution.

File: Dockerfile
Comment (mpdude) [reply Dockerfile:89]: Done in #5667

File: Dockerfile
Comment (mpdude) [reply Dockerfile:57]: Done in #5667

File: Dockerfile
Comment (mpdude) [reply Dockerfile:79]: Done in #5667

**Comment by mpdude** [comment]
I think #5667 solved the original comment.

So – shall we create other/new issues for the questions:
- Can the "scoped" build process remain as-is (without Docker)?
- Is it worthwhile to use the same build script when building the "scoped" variant locally, in GHA and for the Docker images?

**Comment by JanMikes** [comment]
Let's create new issue/pr please, this is getting too messy and i believe this issue was solved via #5667 

> Can the "scoped" build process remain as-is (without Docker)?
> Is it worthwhile to use the same build script when building the "scoped" variant locally, in GHA and for the Docker images?>

From what Tomas said, i believe if the process will be working and will bring some benefits (etc no code duplication, simplicity, performance,...) there is no reason to reject it 😊

**Pull request closed without merging**
